pkg/registry/common/expire/nse_server.go

```diff
@@ -20,71 +20,128 @@ import (
 	"context"
 	"time"
 
+	"github.com/edwarnicke/serialize"
 	"github.com/golang/protobuf/ptypes/empty"
-	"github.com/networkservicemesh/api/pkg/api/registry"
 	"google.golang.org/protobuf/types/known/timestamppb"
 
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
 	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
-	"github.com/networkservicemesh/sdk/pkg/tools/extend"
 )
 
 type expireNSEServer struct {
-	timers        timerMap
+	ctx           context.Context
 	nseExpiration time.Duration
+	timers        unregisterTimerMap
+}
+
+type unregisterTimer struct {
+	expirationTime    time.Time
+	started, canceled bool
+	timer             *time.Timer
+	executor          serialize.Executor
+}
+
+// NewNetworkServiceEndpointRegistryServer wraps passed NetworkServiceEndpointRegistryServer and monitor Network service endpoints
+func NewNetworkServiceEndpointRegistryServer(ctx context.Context, nseExpiration time.Duration) registry.NetworkServiceEndpointRegistryServer {
+	return &expireNSEServer{
+		ctx:           ctx,
+		nseExpiration: nseExpiration,
+	}
 }
 
 func (n *expireNSEServer) Register(ctx context.Context, nse *registry.NetworkServiceEndpoint) (*registry.NetworkServiceEndpoint, error) {
+	t, loaded := n.timers.LoadAndDelete(nse.Name)
+	stopped := loaded && t.timer.Stop()
+
+	var expirationTime time.Time
+	var started bool
+	if stopped {
+		expirationTime = t.expirationTime
+	} else if loaded {
+		<-t.executor.AsyncExec(func() {
+			started = t.started
+			if !started {
+				t.canceled = true
+			}
+		})
+	}
+
 	resp, err := next.NetworkServiceEndpointRegistryServer(ctx).Register(ctx, nse)
 	if err != nil {
+		if stopped {
+			// Timer has been stopped, we need only to reset it.
+			t.timer.Reset(time.Until(expirationTime))
+		} else if loaded && !started {
+			// Timer function has been stopped with the `canceled` flag, we need to remove the flag.
+			t.executor.AsyncExec(func() {
+				t.canceled = false
+				if t.started {
+					// Timer function has been already finished, we need to reset the timer.
+					t.timer.Reset(time.Until(expirationTime))
```

Comment:
expirationTime is not defined for this branch.

Reply:
Yep, replaced with `0`.